runtime/compiler/z/codegen/J9S390CHelperLinkage.cpp

```diff
@@ -439,3 +459,537 @@ TR::Register * TR::S390CHelperLinkage::buildDirectDispatch(TR::Node * callNode,
+         }
+      else
+         {
+         cursor = generateRXInstruction(cg(), TR::InstOpCode::getStoreOpCode(), callNode, javaStackPointerRegister,
+                                        generateS390MemoryReference(vmThreadRegister, offsetJ9SP, cg()));
+         }
+
+#if defined(J9ZOS390)
+      padding += 2;
+      // Loading DSAPointer Register
+
+      if (isJNIIFAHelper)
+         {
+         // JNI IFA helpers before and after the JNI call should use the same system SP as the JNI call itself.
+         DSAPointerReg = systemStackPointer;
+         }
+      else
+         DSAPointerReg = realRegisterManager.use(getDSAPointerRegister());
+
+      generateRXInstruction(cg(), TR::InstOpCode::getLoadOpCode(), callNode, DSAPointerReg,
+                            generateS390MemoryReference(vmThreadRegister, offsetOfSSP, cg()));
+
+      generateSS1Instruction(cg(), TR::InstOpCode::XC, callNode, pointerSize,
+                             generateS390MemoryReference(vmThreadRegister, offsetOfSSP, cg()),
+                             generateS390MemoryReference(vmThreadRegister, offsetOfSSP, cg()));
+
+      static bool saveRegsForPostJniIfaCall = feGetEnv("jni_save_regs_post_ifa") != NULL;
+
+      if (isJNIIFAHelper)
+         {
+         if (isPreJNIHelper                                             // is pre jni IFA
+                 || (saveRegsForPostJniIfaCall && !isPreJNIHelper)) // is post jni ifa and enable reg saving for post jni ifa call
+            saveRegistersForIFAHelper(callNode, systemStackPointer, vmThreadRegister);
+         }
+
+#if defined(TR_HOST_32BIT)
+      padding += 2;
+      // Loading CAA Pointer Register
+      TR::Register *CAAPointerReg = realRegisterManager.use(getCAAPointerRegister());
+      TR_J9VMBase *fej9 = (TR_J9VMBase *) cg()->comp()->fe();
+      int32_t J9TR_CAA_save_offset = fej9->getCAASaveOffset();
+      generateRXInstruction(cg(), TR::InstOpCode::getLoadOpCode(), callNode, CAAPointerReg,
+                            generateS390MemoryReference(DSAPointerReg, J9TR_CAA_save_offset, cg()));
+#endif
+#endif
+      }
+
+   TR::Register* nativeMethodEntryAddressReg = realRegisterManager.use(TR::RealRegister::GPR6);
+
+   // get JavaVM from VMThread
+   cursor = generateRXInstruction(cg(), TR::InstOpCode::getLoadOpCode(), callNode, nativeMethodEntryAddressReg,
+                         generateS390MemoryReference(vmThreadRegister,
+                                                     offsetof(J9VMThread, javaVM), cg()));
+
+
+   // get jitConfig from JavaJVM
+   cursor = generateRXInstruction(cg(), TR::InstOpCode::getLoadOpCode(), callNode, nativeMethodEntryAddressReg,
+                         generateS390MemoryReference(nativeMethodEntryAddressReg,
+                                                     offsetof(J9JavaVM, jitConfig), cg()));
+
+
+   // get J9TR_JitConfig_fast_jitPreJNICallOffloadCheck from jit config
```

Comment:
Isn't this a compile time constant? Why do we need these 3 indirections? i.e. how is this any different from `fast_jitInstanceOf`? Are we able to extract the address statically at compile time via:

```
callNode->getSymbolReference()->getSymbol()->castToMethodSymbol()->getMethodAddress();
```

Reply:
Yes, they are compile time constants. The indirections are copied from the old glue code. 

I had difficulty building direct `BRASL` calls to `fast_jitPre(post)JNICallOffloadCheck`. VM changes might be needed for us to jump directly to `fast_jitPostJNICallOffloadCheck`.